[WPE] Register inspector resources in Internals::openDummyInspectorFrontend

fujii opened the pull request:
#### 620fdedb5d0552fb06883a34ea7b442c50a0a111
<pre>
[WPE] Register inspector resources in Internals::openDummyInspectorFrontend
<a href="https://bugs.webkit.org/show_bug.cgi?id=298086">https://bugs.webkit.org/show_bug.cgi?id=298086</a>

Reviewed by Carlos Garcia Campos.

Some inspector tests were timing out because
window.internals.openDummyInspectorFrontend() failed to load the
inspector resource. Here is the error message:

&gt; Failed to load &apos;resource:///org/webkit/inspector/UserInterface/TestStub.html&apos;.

The inspector resources wasn&apos;t available for the dummy inspector
frontend window because it wasn&apos;t registered by calling
g_resources_register when the window was loading the resource.
Registered the inspector resources in Internals::openDummyInspectorFrontend.

Extracted code to register the inspector resources as a funtion
WTF::registerInspectorResourceIfNeeded().

* Source/JavaScriptCore/PlatformWPE.cmake:
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
(Inspector::backendCommands):
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/glib/GResources.cpp: Added.
(WTF::registerInspectorResourceIfNeeded):
* Source/WTF/wtf/glib/GResources.h: Added.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::openDummyInspectorFrontend):
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/WebProcess/Inspector/wpe/WebInspectorUIWPE.cpp:
(WebKit::WebInspectorUI::didEstablishConnection):
* Tools/CISupport/built-product-archive:
(archiveBuiltProduct):

Canonical link: <a href="https://commits.webkit.org/299647@main">https://commits.webkit.org/299647@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22630bb9a757509ca94cf64ffd3ab8fc7c4506f6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119612 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39306 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29958 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125891 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71686 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121488 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40002 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47884 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90850 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60151 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31930 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107251 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71365 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30966 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25355 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69541 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/111739 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101396 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25548 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128860 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/118130 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46534 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35248 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99447 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46899 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103444 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99289 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44716 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22735 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43098 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19044 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46396 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52102 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146828 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45862 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37738 "Found 1 new JSC binary failure: testapi, Found 18679 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default, ChakraCore.yaml/ChakraCore/test/Function/apply.js.default, ChakraCore.yaml/ChakraCore/test/Function/apply3.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49211 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47548 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WebProcess/wpe/WebProcessMainWPE.cpp:
```diff
@@ -56,6 +56,22 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+    GRefPtr<GResource> gresource = adoptGRef(g_resource_load(gResourceFilename.get(), &error.outPtr()));
+    if (!gresource)
+        g_error("Error loading inspector.gresource: %s", error->message);
+    g_resources_register(gresource.get());
```
I don't remember why we have this code duplicated, but I think the idea in WPE is that we don't want to load the resources unless the inspector is actually used.

aperezdc replied:
This is exactly the reason. Maybe in `DEVELOPER_MODE` builds we can force loading the resources earlier, and keep loading them lazily in non-developer builds. If anyone has a better idea, I'm all ears 😄

aperezdc replied:
Factoring out the duplicated code into a function is a good idea, though. It would be nice to land the refactoring at least.

fujii replied:
Fixed.

fujii commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -267,6 +271,9 @@ void WebLoaderStrategy::scheduleLoad(ResourceLoader& resourceLoader, CachedResou
         LOG(NetworkScheduling, "(WebProcess) WebLoaderStrategy::scheduleLoad, url '%s' will be handled as a GResource.", resourceLoader.url().string().utf8().data());
         WEBLOADERSTRATEGY_RELEASE_LOG("scheduleLoad: URL will be handled as a GResource");
+#if PLATFORM(WPE)
+        WTF::registerInspectorResources();
```
Should I check if the path starts with "/org/webkit/inspector" here?

carlosgcampos commented on Source/WTF/wtf/glib/GResources.cpp:
```diff
@@ -0,0 +1,57 @@
+
+#if PLATFORM(WPE)
+
+void registerInspectorResources()
```
Since this is called from multiple places unconditionally, and I think it can be called more than once, I would call this registerInspectorResourceIfNeeded

fujii replied:
Fixed.

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/glib/InjectedBundleGlib.cpp:
```diff
@@ -49,6 +53,11 @@ bool InjectedBundle::initialize(const WebProcessCreationParameters&, RefPtr<API:
+
+#if PLATFORM(WPE)
+    WTF::registerInspectorResources();
+#endif
```
I don't understand why doing it here, it means that any browser using injected bundle will load the resources unconditionally. Why don't we call this from Internals::openDummyInspectorFrontend?

fujii replied:
Oops. I wanted to do it in the injected bundle of WebKitTestRunner. I'm not confident the web process of inspector is always same with the inspected window. It works in openDummyInspectorFrontend. Will do that.

fujii replied:
Fixed.

fujii committed:
[WPE] Register inspector resources in Internals::openDummyInspectorFrontend
https://bugs.webkit.org/show_bug.cgi?id=298086

Reviewed by Carlos Garcia Campos.

Some inspector tests were timing out because
window.internals.openDummyInspectorFrontend() failed to load the
inspector resource. Here is the error message:

> Failed to load 'resource:///org/webkit/inspector/UserInterface/TestStub.html'.

The inspector resources wasn't available for the dummy inspector
frontend window because it wasn't registered by calling
g_resources_register when the window was loading the resource.
Registered the inspector resources in Internals::openDummyInspectorFrontend.

Extracted code to register the inspector resources as a funtion
WTF::registerInspectorResourceIfNeeded().

* Source/JavaScriptCore/PlatformWPE.cmake:
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorUtils.cpp:
(Inspector::backendCommands):
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/glib/GResources.cpp: Added.
(WTF::registerInspectorResourceIfNeeded):
* Source/WTF/wtf/glib/GResources.h: Added.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::openDummyInspectorFrontend):
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/WebProcess/Inspector/wpe/WebInspectorUIWPE.cpp:
(WebKit::WebInspectorUI::didEstablishConnection):
* Tools/CISupport/built-product-archive:
(archiveBuiltProduct):

Canonical link: https://commits.webkit.org/299647@main
